Kubevirt VmTemplateList

yaacov opened the pull request:
Introduces List Page for Virtual Machines.

![OKD](https://user-images.githubusercontent.com/2181522/59753456-d9489480-928c-11e9-94bc-beafa6ea3c3a.png)

![OKD(1)](https://user-images.githubusercontent.com/2181522/59754186-4a3c7c00-928e-11e9-8fa7-ae461ff83261.png)

Override template yaml:
![OKD(2)](https://user-images.githubusercontent.com/2181522/59754207-558fa780-928e-11e9-9368-b422470ad0fc.png)

yaacov commented:
@mareklibra @vojtechszocs @rawagner original PR #1637 is closed ... continue here.

yaacov commented:
@mareklibra @vojtechszocs @rawagner using page route seems to work :-) please review.

yaacov commented:
Removed WIP and removed dev dependency for kubvirt plugin.

mareklibra commented on frontend/packages/kubevirt-plugin/src/plugin.ts:
```diff
@@ -65,6 +79,26 @@ const plugin: Plugin<ConsumedExtensions> = [
       template: yamlTemplates.getIn([models.VirtualMachineModel, 'default']),
```
Shouldn't be yaml template customized as well?

yaacov replied:
Yes, I forgot it :sunglasses: , added.

yaacov replied:
Note, when defining a template we override the `TemplateModel` yaml template, so if we want to support a different yaml for templates we will have to write a new page renderer for `~new`.

vojtechszocs replied:
> Note, when defining a template we override the `TemplateModel` yaml template

YAML templates are keyed by model & template name. In this case, we should do something like:

```ts
{
  type: 'YAMLTemplate',
  properties: {
    model: TemplateModel,
    template: 'VM template YAML goes here',
    templateName: 'vm-template',
  },
},
```

yaacov replied:
Nice, thanks !

mareklibra commented on frontend/packages/kubevirt-plugin/src/plugin.ts:
```diff
@@ -65,6 +79,26 @@ const plugin: Plugin<ConsumedExtensions> = [
+  {
+    type: 'Page/Route',
+    properties: {
+      path: `/k8s/ns/:ns/${models.VmTemplateModel.plural}`,
```
Is this needed? We have a resource list page and specialized model. I think generic router rule should work. Please correct me.

yaacov replied:
This is what I understood from https://github.com/openshift/console/pull/1658#issuecomment-499574443 , I may have misunderstood the discussion ...

**Background**
AFAIU `VmTemplateModel` is of kind `Template` so defining it will redefining the existing model `Template`, so `ModelDefinition` skips it. Once I added the template we actually get a console warning:
```
attempt to redefine model template.openshift.io~v1~Template
```

vojtechszocs replied:
Yes, KubeVirt reuses the existing `Template` resource (`template.openshift.io~v1~Template`).

There's no need to redefine that model. What we need is a resource list page which applies a default filter to show VM templates only, e.g. via `Table.staticFilters` prop.

vojtechszocs replied:
> We have a resource list page and specialized model.

We don't have a specialized model for VM template, we reuse existing one.

Mapping between models vs. resource list/details pages is 1-1 so we need to use route-based page here.

yaacov commented on frontend/packages/kubevirt-plugin/src/components/vm-template.tsx:
```diff
@@ -0,0 +1,112 @@
+export const VmTemplatesPageTitle = 'Virtual Machine Templates';
+
+const createItems = {
+  wizard: 'Create with Wizard',
```
Removing :+1:

yaacov commented on frontend/packages/kubevirt-plugin/src/components/vm-template.tsx:
```diff
@@ -0,0 +1,112 @@
+
+import {
+  getTemplateOperatingSystems,
+  DASHES,
```
Replaced :hear_no_evil:

spadgett commented:
This needs to use PF4 tables. See #1465

yaacov commented:
wip until pf4 table integration finished ... :-(

yaacov commented:
> This needs to use PF4 tables. See #1465

Remove wip, :cake: converted to tables ... 

@spadgett  @mareklibra  please review.

yaacov commented:
Added details page :tada: 

![Peek 2019-06-18 15-51](https://user-images.githubusercontent.com/2181522/59683604-24a56900-91e1-11e9-8169-b3a749ac2527.gif)

rawagner commented:
@yaacov can you please use the latest `web-ui-components` which has updated Details page design

vojtechszocs commented:
@yaacov The scope of this PR is to add both VM template list and details pages, right?

vojtechszocs commented:
> @yaacov can you please use the latest `web-ui-components` which has updated Details page design

This relates to #1738, which should get merged soon. Please rebase on top of that.

yaacov commented:
> @yaacov The scope of this PR is to add both VM template list and details pages, right?

No, I plan to move the details to a different PR ... so this one will be only about the table.

yaacov commented:
> This relates to #1738, which should get merged soon. Please rebase on top of that.

Yes :+1: , @vojtechszocs thanks, I will need to sync with @mareklibra on that.

mareklibra commented:
@yaacov , latest `web-ui-components` are in place now (#1738 is merged).

When looking at the screencast above, the Details page should be re-designed in similar way as in #1682.

yaacov commented:
@spadgett @vojtechszocs @mareklibra please review.

Changes from last review cycle :-)

  - use `components/factory` `Table`
  - use `Page/Route` ( vmtemplate is not a resource, can't use `Page/Resource/List` )
  - VmTemplate specific code moved to a separate dir like #1682
  - Overrides Template model yaml page

yaacov commented:
/test e2e-aws-console-olm

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vmtemplates/vm-template.tsx:
```diff
@@ -0,0 +1,123 @@
+      <TableData className={tableColumnClasses[0]}>
+        <ResourceLink
+          kind={kind}
+          name={template.metadata.name}
```
there are selectors in `@console/shared`

yaacov replied:
Thanks !

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vmtemplates/vm-template.tsx:
```diff
@@ -0,0 +1,123 @@
+          kind={kind}
+          name={template.metadata.name}
+          namespace={template.metadata.namespace}
+          title={template.metadata.uid}
```
this one is still missing there

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vmtemplates/vm-template.tsx:
```diff
@@ -0,0 +1,123 @@
+      title: 'Namespace',
+      sortField: 'metadata.namespace',
+      transforms: [sortable],
+      props: { className: tableColumnClasses[0] },
```
IMO it would be better to map the same index of tableColumnClasses to the Header index

- it is easier to read and make changes to

yaacov replied:
Ahhh ... wow, I saw it in the example ... and didn't understand what i saw :blush: 

Thanks, will fix that  :exclamation:

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vmtemplates/vm-template.tsx:
```diff
@@ -0,0 +1,123 @@
+const { kind, selector } = VmTemplateModel;
+
+const tableColumnClasses = [
+  classNames('col-lg-2', 'col-sm-4', 'col-xs-4'),
```
we could make constants out of these if we reuse them for multiple indexes

yaacov replied:
Make sense to me too :+1:

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vmtemplates/vm-template.tsx:
```diff
@@ -0,0 +1,123 @@
+        />
+      </TableData>
+      <TableData className={tableColumnClasses[0]}>
+        {_.get(template.metadata, 'annotations.description', '-')}
```
there is a DASH constant already in `@console/shared`

yaacov replied:
Thanks !

yaacov commented:
@spadgett @vojtechszocs @mareklibra @suomiy please review.

Addressed issues raised by @suomiy and  @spadgett in #1754 

Changes from last review cycle:

-  Addressed issues raised by @suomiy 
-  Removed `VmTemplateModel` model ( because it is not a model, just `TemplatModel` with selector ) 
-  Removed the override of template yaml page
-  Chenaged `vmtemplate` link from resource link to href link.

yaacov commented:
wip, href link does not work as expected, looking for ways to fix :-(

yaacov commented:
removed wip, and restored the resource link with a comment explaining that vmtemplate is actually a resource in this context ( e.g. it's a template with a selector )

yaacov commented:
@spadgett hi, I tried to address https://github.com/openshift/console/pull/1754#discussion_r295429625 but it looks to me that this nav item is actually a link to a resource, `template + selector`, and need to use the resource link, WDYT ?

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/components/vmtemplates/vmtemplates.tsx:
```diff
@@ -0,0 +1,136 @@
+    aria-label={labelPlural}
+    Header={VmTempleateTableHeader}
+    Row={VmTemplateTableRow}
+    virtualize
```
Please note, the `virtualize` prop should default to `true` (based on discussion with @priley86).

yaacov replied:
:+1:

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/components/vmtemplates/vmtemplates.tsx:
```diff
@@ -0,0 +1,136 @@
+    title={labelPlural}
+    canCreate
+    ListComponent={VirtualMachineTemplates}
+    {...props}
```
Nit: the standard convention is to put `props` spread at the top, ensuring that they don't override our explicit props unexpectedly:

```tsx
<ListPage
  {...props}
  title={labelPlural}
  canCreate
  ListComponent={VirtualMachineTemplates}
  kind={kind}
  selector={selector}
/>
```

yaacov replied:
Thanks :white_flower:

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/models.ts:
```diff
@@ -48,24 +48,6 @@ export const VirtualMachineInstanceReplicaSetModel: K8sKind = {
 };
 
-/* TODO(mlibra): migrate templates
-export const VmTemplateModel: K8sKind = {
```
:+1:

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/plugin.ts:
```diff
@@ -50,6 +52,21 @@ const plugin: Plugin<ConsumedExtensions> = [
     },
   },
+  {
+    // NOTE(yaacov): vmtemplates is a template resource with a slector.
```
Nit: misspelled "selector".

yaacov replied:
Thanks :blush:

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/components/vmtemplates/vmtemplates.tsx:
```diff
@@ -0,0 +1,136 @@
+import * as React from 'react';
```
Nit: please use `src/components/vm-templates` directory and `vm-template.tsx` module to stick with existing name conventions.

yaacov replied:
Cool :+1:

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/plugin.ts:
```diff
@@ -50,6 +53,21 @@ const plugin: Plugin<ConsumedExtensions> = [
+  {
+    // NOTE(yaacov): vmtemplates is a template resource with a selector.
+    // 'NavItem/ResourceNS' is used, and not 'NavItem/Href', because it injects
+    // the namespace needed to get the currect link to a resource ( template with selector ) in our case.
```
"correct link"

yaacov replied:
ahhh

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/plugin.ts:
```diff
@@ -77,6 +103,26 @@ const plugin: Plugin<ConsumedExtensions> = [
+  {
+    type: 'Page/Route',
+    properties: {
+      path: `/k8s/all-namespaces/vmtemplates`,
```
If you need multiple paths associated with the same page component, you can do this:

```ts
{
  type: 'Page/Route',
  properties: {
    path: ['/k8s/all-namespaces/vmtemplates', '/k8s/ns/:ns/vmtemplates'],
    loader: () => {/* dynamically load VirtualMachineTemplatesPage */},
  },
},
```

There's no need to use string literals (back-ticks) if the value doesn't contain any expressions, using `'` for static strings is enough.

Also, consider adding `exact: true` to denote that we want an exact URL match.

Please check out the DevConsole plugin (`packages/dev-console/src/plugin.tsx`) and how it uses `Page/Route` extensions.

yaacov replied:
Thanks !

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/yaml-templates.ts:
```diff
@@ -39,4 +41,77 @@ spec:
+  )
+  .setIn(
+    [TemplateModel, 'vm-template'],
+
```
Nit: extra newline.

yaacov replied:
sorry :-)

yaacov commented:
/test e2e-aws-console-olm

yaacov commented:
@spadgett please review

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-templates/vm-template.tsx:
```diff
@@ -0,0 +1,135 @@
+};
+VmTempleateTableHeader.displayName = 'VmTemplateTableHeader';
+
+const VmTemplateTableRow = ({ obj: template, index, key, style }) => {
```
please, can we use the types as follows?

```ts
VmTemplateTableRow: React.FC<VmTemplateTableRowProps>
```

yaacov replied:
sure :+1:

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-templates/vm-template.tsx:
```diff
@@ -0,0 +1,135 @@
+          kind={kind}
+          name={getName(template)}
+          namespace={getNamespace(template)}
+          title={template.metadata.uid}
```
`getUid` is available from `@console/shared` as of today.

yaacov replied:
yeee :-)

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-templates/vm-template.tsx:
```diff
@@ -0,0 +1,135 @@
+};
+VmTemplateTableRow.displayName = 'VmTemplateTableRow';
+
+const VirtualMachineTemplates = (props) => (
```
```ts
VirtualMachineTemplates: React.FC<React.ComponentProps<typeof Table> & VirtualMachineTemplatesProps>
```

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-templates/vm-template.tsx:
```diff
@@ -0,0 +1,135 @@
+    {...props}
+    aria-label={labelPlural}
+    Header={VmTempleateTableHeader}
+    Row={VmTemplateTableRow}
```
I think we should be using `virtualize` for the main tables

yaacov replied:
in https://github.com/openshift/console/pull/1673#discussion_r295886830 

@vojtechszocs I removed it because it's true by default ... did I miss something ?

suomiy replied:
ah okay, you are right. I missed that

yaacov replied:
np, thanks for looking :+1:

vojtechszocs replied:
@suomiy The `virtualize` prop should always default to `true` in Console table components.

In both `Table(Inner).render` and `List(Inner).render` methods, there's `virtualize = true` destructuring from `props`. In other words, all Console tables should be rendered as "virtualized" by default.

vojtechszocs replied:
I'm OK with having an explicit `virtualized` prop here, just tried to say is that its default value is `true` anyway.

suomiy commented on frontend/packages/kubevirt-plugin/src/yaml-templates.ts:
```diff
@@ -1,10 +1,12 @@
-export const yamlTemplates = ImmutableMap().setIn(
-  [VirtualMachineModel, 'default'],
-  `
+export const yamlTemplates = ImmutableMap()
```
can we move this file to a special folder to not to polute the `src`?

yaacov replied:
@suomiy this yaml const is used by vms, vm-templates and in the future, all other resources.

I agree about moving this file, because moving it impacts not only this PR but also existing `vms` resources and the future resources we will import from web-ui, it should be done in a separate PR and inform all people working on the that we intend to move this file.

suomiy replied:
ok it can be done later if you think this is a hassle (even though the conflicts should be resolvable in each PR)

vojtechszocs replied:
+1 on that.

I'd suggest to create `src/models` and `git mv src/models.ts src/models/index.ts`.

YAML templates are inherently related to models, so I'd add `src/models/yaml-templates.ts`.

yaacov commented:
@suomiy thanks :+1: addressed the comments except the one about moving the yaml templates, It will impact orher resources and IMHO should be done separate from this PR.

p.s.
squashed commits because it became more then 20 commits ... :-)

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-templates/vm-template.tsx:
```diff
@@ -0,0 +1,149 @@
+);
+
+type VmTemplateTableRowProps = {
+  template: K8sResourceKind;
```
TemplateKind should be more concrete

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-templates/vm-template.tsx:
```diff
@@ -0,0 +1,149 @@
+  key: string;
+  style: any;
+};
+type VirtualMachineTemplatesProps = {};
```
having empty types is not much useful

yaacov replied:
:sunglasses:

yaacov commented on frontend/packages/kubevirt-plugin/src/components/vm-templates/vm-template.tsx:
```diff
@@ -0,0 +1,149 @@
+};
+VmTempleateTableHeader.displayName = 'VmTemplateTableHeader';
+
+const VmTemplateTableRow = ({ template, index, key, style }: VmTemplateTableRowProps) => {
```
:smile: :+1:

yaacov commented:
@suomiy thanks for the comments :+1: ,  broke some things in the way, now it's working again :-)

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/components/vm-templates/vm-template.tsx:
```diff
@@ -0,0 +1,145 @@
+  Kebab.columnClass,
+];
+
+const VmTempleateTableHeader = () => {
```
Typo: `VmTemplateTableHeader`.

yaacov replied:
Thanks :sunglasses:

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/plugin.ts:
```diff
@@ -50,6 +53,21 @@ const plugin: Plugin<ConsumedExtensions> = [
+  {
+    // NOTE(yaacov): vmtemplates is a template resource with a selector.
+    // 'NavItem/ResourceNS' is used, and not 'NavItem/Href', because it injects
+    // the namespace needed to get the correct link to a resource ( template with selector ) in our case.
```
In theory, `NavItem/Href` should only be used for URLs outside of Console :smile:

So using `NavItem/ResourceNS` pointing to VM template resource is self-explanatory, I think.

yaacov replied:
This comment is here because of https://github.com/openshift/console/pull/1754#discussion_r295429625 

May be an over kill :smile: I can remove it if you go not think someone may refactor it into an href in the future :tulip:

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/plugin.ts:
```diff
@@ -77,6 +103,17 @@ const plugin: Plugin<ConsumedExtensions> = [
+    type: 'Page/Route',
+    properties: {
+      exact: true,
+      path: [`/k8s/ns/:ns/vmtemplates`, '/k8s/all-namespaces/vmtemplates'],
```
Nit: no need for `` ` `` in static strings.

yaacov replied:
:+1:

vojtechszocs commented on frontend/packages/kubevirt-plugin/src/yaml-templates.ts:
```diff
@@ -39,4 +41,76 @@ spec:
+    metadata:
+      creationTimestamp: null
+      labels:
+        kubevirt-vm: 'vm-\${NAME}'
```
I'm probably missing something but where does `NAME` (and other uppercase-labeled variables) come from?

yaacov replied:
It's the template pramters, https://docs.okd.io/latest/dev_guide/templates.html#writing-parameters

default values defined here:
https://github.com/openshift/console/pull/1673/files#diff-a05f509d407deb7b65bb87cd43944154R106

vojtechszocs replied:
@yaacov Thank you for clarification.

vojtechszocs commented:
> squashed commits because it became more then 20 commits ... :-)

When working with multiple un-merged PRs stacked onto each other, it's more practical & easier to review (in my opinion) when one PR is one commit.

yaacov commented:
@vojtechszocs Thanks ^^

yaacov commented:
@vojtechszocs @suomiy please review

yaacov commented:
@suomiy thanks ! this PR also need lgtm label :tractor:

suomiy commented:
/lgtm
